folding-schemes/src/folding/hypernova/mod.rs

```diff
@@ -25,3 +55,534 @@ impl<F: PrimeField> Witness<F> {
+    pub ccs: CCS<C1::ScalarField>,
+    pub cf_r1cs: R1CS<C2::ScalarField>,
+    pub cs_params: CS1::ProverParams,
+    pub cf_cs_params: CS2::ProverParams,
+}
+
+/// Implements HyperNova+CycleFold's IVC, described in
+/// [HyperNova](https://eprint.iacr.org/2023/573.pdf) and
+/// [CycleFold](https://eprint.iacr.org/2023/1192.pdf), following the FoldingScheme trait
+#[derive(Clone, Debug)]
+pub struct HyperNova<C1, GC1, C2, GC2, FC, CS1, CS2>
+where
+    C1: CurveGroup,
+    GC1: CurveVar<C1, CF2<C1>> + ToConstraintFieldGadget<CF2<C1>>,
+    C2: CurveGroup,
+    GC2: CurveVar<C2, CF2<C2>>,
+    FC: FCircuit<C1::ScalarField>,
+    CS1: CommitmentScheme<C1>,
+    CS2: CommitmentScheme<C2>,
+{
+    _gc1: PhantomData<GC1>,
+    _c2: PhantomData<C2>,
+    _gc2: PhantomData<GC2>,
+
+    /// CCS of the Augmented Function circuit
+    pub ccs: CCS<C1::ScalarField>,
+    /// R1CS of the CycleFold circuit
+    pub cf_r1cs: R1CS<C2::ScalarField>,
+    pub poseidon_config: PoseidonConfig<C1::ScalarField>,
+    /// CommitmentScheme::ProverParams over C1
+    pub cs_params: CS1::ProverParams,
+    /// CycleFold CommitmentScheme::ProverParams, over C2
+    pub cf_cs_params: CS2::ProverParams,
+    /// F circuit, the circuit that is being folded
+    pub F: FC,
+    pub i: C1::ScalarField,
+    /// initial state
+    pub z_0: Vec<C1::ScalarField>,
+    /// current i-th state
+    pub z_i: Vec<C1::ScalarField>,
+    /// HyperNova instances
+    pub W_i: Witness<C1::ScalarField>,
+    pub U_i: LCCCS<C1>,
+    pub w_i: Witness<C1::ScalarField>,
+    pub u_i: CCCS<C1>,
+
+    /// CycleFold running instance
+    pub cf_W_i: NovaWitness<C2>,
+    pub cf_U_i: CommittedInstance<C2>,
+}
+
+impl<C1, GC1, C2, GC2, FC, CS1, CS2> FoldingScheme<C1, C2, FC>
+    for HyperNova<C1, GC1, C2, GC2, FC, CS1, CS2>
+where
+    C1: CurveGroup,
+    GC1: CurveVar<C1, CF2<C1>> + ToConstraintFieldGadget<CF2<C1>>,
+    C2: CurveGroup,
+    GC2: CurveVar<C2, CF2<C2>> + ToConstraintFieldGadget<CF2<C2>>,
+    FC: FCircuit<C1::ScalarField>,
+    CS1: CommitmentScheme<C1>,
+    CS2: CommitmentScheme<C2>,
+    <C1 as CurveGroup>::BaseField: PrimeField,
+    <C2 as CurveGroup>::BaseField: PrimeField,
+    <C1 as Group>::ScalarField: Absorb,
+    <C2 as Group>::ScalarField: Absorb,
+    C1: CurveGroup<BaseField = C2::ScalarField, ScalarField = C2::BaseField>,
+    for<'a> &'a GC1: GroupOpsBounds<'a, C1, GC1>,
+    for<'a> &'a GC2: GroupOpsBounds<'a, C2, GC2>,
+{
+    type PreprocessorParam = PreprocessorParam<C1, C2, FC, CS1, CS2>;
+    type ProverParam = ProverParams<C1, C2, CS1, CS2>;
+    type VerifierParam = VerifierParams<C1, C2, CS1, CS2>;
+    type RunningCommittedInstanceWithWitness = (LCCCS<C1>, Witness<C1::ScalarField>);
+    type IncomingCommittedInstanceWithWitness = (CCCS<C1>, Witness<C1::ScalarField>);
+    type CFCommittedInstanceWithWitness = (CommittedInstance<C2>, NovaWitness<C2>);
+
+    fn preprocess(
+        mut rng: impl RngCore,
```

Comment:
I'm not sure (due my ignorance) about this one, because we're using the arkworks test_rng generator which allows us to use a deterministic 'randomness' for the tests, which is very useful, whereas if we require CryptoRng then that can not be used :S

Reply:
Hmmm I see. Usually `CryptoRng` is a required trait for most of the rng-providers that have to do with cryptography since it marks the generator as cryptographically secure.

Not sure how they handle this in `arkworks`. 
But you can read more about this here: https://rust-random.github.io/rand/rand_core/trait.CryptoRng.html